language/module.md

```diff
@@ -0,0 +1,71 @@
+# Puppet Module Schema
+
+This document describes allowed (optional) and required files and directories in
+a Puppet Module.
+
+## What is a Puppet Module
+
+A module in the sense of this document is a compressed tar archive. It is
+usually distributed via [forge.puppet.com](https://forge.puppet.com/). A module
+is usually developed in a git repository. Some files in the git repository are
+only used for development and testing. They shall not be released.
+
+Common files often seen in a vcs repository that are used for development but
+shall not be released:
+
+* `/spec/`
+* `/Rakefile`
+* `/Gemfile`
+* `/.gitignore`
+* `/.github/`
+* `/.devcontainer`
+* `/Dockerfile`
+
+Note that above are just examples and not a complete list. The goal of this
+document is to provide an allowlist, not a denylist.
+
+The official
+[Puppet documentation](https://www.puppet.com/docs/puppet/latest/modules_fundamentals.html)
+already explains what a module is and what it can contain.
+
+| Directories and Files | Purpose |
+|-----------------------|---------|
+| `/manifests/`         | contains Puppet code |
+| `hiera.yaml`          | Can define a Hiera configuration for Hiera data within this module |
+| `/data/`              | If the module has a `hiera.yaml`, the related data has to be within `/data/` |
```

Comment:
The `datadir` key in `hiera.yaml` can be set to anything. Would this also imply a modification in Puppet to enforce this? If so, should the `defaults` be preset?

Reply:
the current default is `/data` when `datadir` isn't set. I would like to see puppet enforcing it.

Reply:
I agree, thihs makes stfuff more complicated. We could change the spec and enforce it in `/data`.

Reply:
Perhaps something that in current Puppet can be deprecated (including warning, plus CI check) and removed in Puppet 9?

Which also implies this document should somehow be versioned. I've been a fan of the Puppet report format documentation with its explicit version numbers and "changes since previous version". Would that be a good concept to copy here?

Reply:
I had that in mind as well, but wasnt sure how to write it since this is the initial document.

Reply:
Puppet's report format was initially version 0. I'd make this version 0, formalizing what's already out there in the world. Then Puppet 9 could formalize version 1.

Reply:
This is a good idea to move forward. I added a section about format and versioning.